Format json option

MartinFillon opened the pull request.

MartinFillon committed:
feat(json): starting to parse json as per asked in #768

MartinFillon committed:
refactor: refactored the way string are created to follow up

MartinFillon committed:
docs(json): changed docs to add new arg

MartinFillon committed:
fix: clippy warnings

MartinFillon committed:
test: regenerated in place tests

MartinFillon committed:
feat(json): updated format to fit requirements

MartinFillon committed:
feat(json): made json available for every output type

MartinFillon commented:
This can either wait for tree tbd or be already merged

PThorpe92 commented:
> This can either wait for tree tbd or be already merged

I have a fix incoming, i know it's going to get piped/formatted most of the time but

PThorpe92 committed:
fix(json): correct brace formatting output for json flag

PThorpe92 commented:
Hey I don't think it still actually equivalent to `--long` view? 

I believe now it requires the flag so we might want to change the docs/completions

MartinFillon commented:
> Hey I don't think it still actually equivalent to --long view?

it is as it just takes every cell used by long view and display its content differently

MartinFillon commented:
> I have a fix incoming, i know it's going to get piped/formatted most of the time but its just changing this

tbh only tested with `jq '.'`

MartinFillon committed:
refactor: moved things around to easy up the tree implementation

I am still having trouble to implement json tree with the current data
If I can't find anything I might refactor a bit of the code to have
a different type of data structure saved to then be printed at once

PThorpe92 committed:
fix: rename variable and remove unneeded fn parameter in details

MartinFillon committed:
feat: first recursive is done need to be able to have depth

MartinFillon committed:
feat: tree is growing

Tree works for small trees but still have problems on big complex trees
such as eza dir.

MartinFillon committed:
refactor: separated json in another file

S1ngularity96 commented:
I'd like to join the conversation for testing and contributing. I tried the `cargo run -- --long --json` command. However the output results in an object, where the key `files` consists of an array of arrays. Is that the right behaviour? Don't we want to have an array with objects that contain keys for each file attribute?

S1ngularity96 commented on src/output/mod.rs:
```diff
@@ -36,6 +38,12 @@ pub enum Mode {
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum OutputType {
+    Legacy,
+    Json,
+}
```
I wouldn't call the default output type `Legacy`. This should be something like Console or TTY?

azaleacolburn replied:
Or just 'Unstructured'.

S1ngularity96 commented on completions/zsh/_eza:
```diff
@@ -69,6 +69,7 @@ __eza() {
         '*:filename:_files' \
         --smart-group"[Only show group if it has a different name from owner]" \
         --stdin"[When piping to eza. Read file names from stdin]"
+        --json"[Output results as JSON(equivalent to --long)]"
```
This will fail. You forgot the `\` character. 

```suggestion
        --smart-group"[Only show group if it has a different name from owner]" \
        --stdin"[When piping to eza. Read file names from stdin]" \
        --json"[Output results as JSON(equivalent to --long)]"
```

S1ngularity96 commented:
I don't like the approach here. I would have changed the codebase so that at the end there would always be a collection of structs with the respective file attributes. An output-specific renderer would then output the content. The code keeps getting bloated if more formats are added. There should be one source file for each format in which the rendering strategy is implemented. The render function always gets the same object so that the interface fits. If possible, other source files should not be changed at all.

Maybe something like this. 

![image](https://github.com/eza-community/eza/assets/16043122/93ec4053-bfa9-4aa2-9236-3f609e87f0fd)

MartinFillon commented:
> I'd like to join the conversation for testing and contributing. I tried the cargo run -- --long --json command. However the output results in an object, where the key files consists of an array of arrays. Is that the right behaviour? Don't we want to have an array with objects that contain keys for each file attribute?

As per the precedent conversations, et to be consistent with current codebase this behaviour needs the `--header`

> Maybe something like this.

I Like the idea you proposing, but that is a bit of work tbh, tho, if you take a look at the current way of doing it, this is already like that as per how I implemented it, its just kinda hidden within the current structs. But refactoring everything is something I am thinking about.

S1ngularity96 commented:
> > I'd like to join the conversation for testing and contributing. I tried the cargo run -- --long --json command. However the output results in an object, where the key files consists of an array of arrays. Is that the right behaviour? Don't we want to have an array with objects that contain keys for each file attribute?
> 
> As per the precedent conversations, et to be consistent with current codebase this behaviour needs the `--header`
> 
> > Maybe something like this.
> 
> I Like the idea you proposing, but that is a bit of work tbh, tho, if you take a look at the current way of doing it, this is already like that as per how I implemented it, its just kinda hidden within the current structs. But refactoring everything is something I am thinking about.

In the future, whenever a new format is added, it will be necessary to refactor it anyway. It would also be better to cleanly separate the data from the rendering logic.
However I also looked through the 'Display Options' again. `Tree`, `Grid` and are all separate formats. It somehow doesn't make sense that it's possible to run `--tree --grid` in one command.

MartinFillon commented:
Well you got a point on the refactor part, as per the options one, there is priority made, but you can use the envvars to be more precise.

MartinFillon commented:
> If you like I could try to reorganize the code and do a pull request. But only if that's okay and its alright for you to change some underlying architecture of the code.

We always up for any contribution go ahead as you wish.

MartinFillon committed:
feat: add recurse option

MartinFillon commented:
fixes #472

MartinFillon committed:
chore: formatted code

MartinFillon commented:
waiting for #787

MartinFillon commented:
Hey @PThorpe92 is this expected that some features of textcell are missing, as the pr no longer builds due to `trim` and `isEmpty` no longer implemented ?

MartinFillon commented:
> Hey @PThorpe92 is this expected that some features of textcell are missing, as the pr no longer builds due to trim and isEmpty no longer implemented ?

up @PThorpe92 ?

PThorpe92 commented:
sorry just saw this, i'll look at it tomorrow. but a bunch of stuff from term_grid is no longer there if thats what u mean. i'll look at this in the morning

MartinFillon commented:
Well if thats intended i can try and reimplement them ?

PThorpe92 commented:
Ok I see what's going on. It's not TextCell it's the AnsiGenericString from `nu_ansi_term`. 

Let me check out nu_ansi_term and see if there's replacement methods

PThorpe92 commented:
Ok so https://docs.rs/nu-ansi-term/latest/nu_ansi_term/struct.AnsiGenericString.html
that's what we are working with. As you an see it's just the `&str` and a `style`. So those methods can be easily replaced. You can access the underlying string with `.as_str()`

MartinFillon committed:
fix: fixed build error

MartinFillon commented:
Okay @cafkafk and @PThorpe92 this is finally ready for merging

MartinFillon commented:
> Since this is hand rolling the json impl it isn't properly escaping invalid characters. For example a filename that has a quote in it produces invalid json.

Oh Thanks didn't thought of that gonna take a look on how to fix it.

PThorpe92 commented:
Have we considered just creating a struct and `impl serde::Serialize` (nu_ansi_term supports this for the ansi codes) and doing `to_string_pretty()`?

MartinFillon commented:
Well, the thing is that the way we are storing everything right now is not really the best, I thought of using this but implementing it gave me more headaches than anything. And I was stuck between refactoring everything just for this to work, or doing this way. As I didn't had the head time to refactor I didn't move with it, but if we take the time to refactor, and stop storing the strings in a vector but a better table that would make it better clearly

nyabinary commented:
Any update to this?

cafkafk reviewed (changes requested):
> Any update to this?

I've lost track to be honest, seems the immediate concern is fixing conflicts if the design is approved.
